File: Monotonic_Stack_29_Oct_2023/0503-next-greater-element-ii.go
package main

func nextGreaterElements(nums []int) []int {
    n := len(nums)
    s := make([]int, n)   
    ans := make([]int, n)
    for i := range ans {
        ans[i] = -1
    }

    for i := 0; i < 2*n; i++ {
        i := i % n
        
        for top := len(s)-1; top >= 0 && nums[s[top]] < nums[i]; top-- {
            ans[s[top]] = nums[i]
            s = s[:top]
        }
        s = append(s, i)
    }

    return ans
}

File: Monotonic_Stack_29_Oct_2023/0503_next_greater_element_ii.py
class Solution:
    def nextGreaterElements(self, nums: List[int]) -> List[int]:
        stack= []
        n = len(nums)
        res = [-1] * n
        
        for i, num in enumerate(nums):
            while stack and nums[stack[-1]] < num:
                res[stack.pop()] = num
            stack.append(i)
        
        for i, num in enumerate(nums):              
            while stack and nums[stack[-1]] < num:
                res[stack.pop()] = num
                
        return res
